src/wp-includes/view-transitions.js

```diff
@@ -0,0 +1,191 @@
+// This script should eventually live elsewhere, but for now it's just in `wp-includes` for simplicity.
+if ( !! window.navigation && 'CSSViewTransitionRule' in window ) {
+	const config = { __PLACEHOLDER__: true };
+
+	const getViewTransitionEntries = ( transitionType, bodyElement, articleElement ) => {
+		const isMainSlide = transitionType === 'forwards' || transitionType === 'backwards';
+		let foundMainElement = false;
+		return [
+			...Object.entries( config.globalTransitionNames || {} ).map( ( [ selector, name ] ) => {
+				const element = bodyElement.querySelector( selector );
+				if ( name === 'main' && element ) {
+					foundMainElement = true;
+				}
+				return [ element, name ];
+			} ),
+			...( articleElement && ( ! isMainSlide || ! foundMainElement )
+				? Object.entries( config.postTransitionNames || {} ).map( ( [ selector, name ] ) => {
+					const element = articleElement.querySelector( selector );
+					return [ element, name ];
+				} )
+				: []
+			),
+		];
+	};
+
+	const setTemporaryViewTransitionNames = async ( entries, vtPromise ) => {
+		for ( const [ element, name ] of entries ) {
+			if ( ! element ) {
+				continue;
+			}
+			element.style.viewTransitionName = name;
+		}
+
+		await vtPromise;
+
+		for ( const [ element, _ ] of entries ) {
+			if ( ! element ) {
+				continue;
+			}
+			element.style.viewTransitionName = '';
+		}
+	};
+
+	const appendSelectors = ( selectors, append ) => {
+		return selectors.split( ',' ).map( subselector => subselector.trim() + ' ' + append ).join( ',' );
+	};
+
+	const getArticle = () => {
+		if ( ! config.postSelector ) {
+			return null;
+		}
+		return document.querySelector( config.postSelector );
+	};
+
+	const getArticleForUrl = ( url ) => {
+		if ( ! config.postSelector ) {
+			return null;
+		}
+		const postLinkSelector = appendSelectors( config.postSelector, 'a[href="' + url + '"]' );
+		const articleLink = document.querySelector( postLinkSelector );
+		if ( ! articleLink ) {
+			return null;
+		}
+		return articleLink.closest( config.postSelector );
+	};
+
+	const determineTransitionType = ( oldEntry, newEntry ) => {
+		if ( ! config.chronologicalSlideInOut ) {
+			return 'default';
+		}
+
+		if ( ! oldEntry || ! newEntry ) {
+			return 'default';
+		}
+
+		const oldURL = new URL( oldEntry.url );
+		const newURL = new URL( newEntry.url );
+
+		const oldPathname = oldURL.pathname;
+		const newPathname = newURL.pathname;
+
+		if ( oldPathname === newPathname ) {
+			return 'default';
+		}
+
+		// Check if the URLs are for a paginated archive.
+		let oldPageMatches = oldPathname.match( /\/page\/(\d+)\/?$/ );
+		let newPageMatches = newPathname.match( /\/page\/(\d+)\/?$/ );
```

Comment:
Technically the `/page/` can be overridden by setting it to something else in `WP_Rewrite::$pagination_base`. So ideally this would be exported from PHP as part of the config.

Reply:
Related, I was thinking last week about whether we should even use the URLs for this or rely on classes on `body`. Almost all relevant WordPress query parameters (including those for pagination) have `body` classes we could use. That would make the implementation more flexible and also work on sites without pretty permalinks.

Part of the JS in this PR already used `body` classes, so maybe we should use that approach more holistically?